plugins/woocommerce/includes/wc-update-functions.php

```diff
@@ -2589,3 +2590,30 @@ function wc_update_770_remove_multichannel_marketing_feature_options() {
 	delete_option( 'woocommerce_multichannel_marketing_enabled' );
 	delete_option( 'woocommerce_marketing_overview_welcome_hidden' );
 }
+
+/**
+ * Delete posts of type "shop_order_placeholder" with no matching order in the orders table.
+ *
+ * @return void
+ */
+function wc_update_790_delete_stray_order_records() {
+	global $wpdb;
+
+	$orders_table_name = OrdersTableDataStore::get_orders_table_name();
+
+	// phpcs:disable WordPress.DB.PreparedSQL
+
+	if ( ! $wpdb->get_var( $wpdb->prepare( 'SHOW TABLES LIKE %s', $wpdb->esc_like( $orders_table_name ) ) ) ) {
+		return;
+	};
+
+	$wpdb->query(
+		$wpdb->prepare(
+			"DELETE FROM {$wpdb->posts} WHERE post_type = %s AND ID NOT IN (SELECT id FROM $orders_table_name)",
```

Comment:
can we batch this, this seems like a pretty big query to do in an action scheduler time limit of 30 seconds. alternatively, can you please do some benchmarking around it so we know that this is not a problem even for large sites.

Reply:
How we would do batching here? I guess we would need to do the `SELECT` separately with a `LIMIT` and with a `WHERE id>x`, and store the highest obtained id to be used in the next batch as the `x`?

Reply:
Yup, something like that would work. We can also use the `BatchProcessorInterface` and perhaps make improvements to it as we go. 

Unless we can test it out on some large site and prove that this isn't a problem that warrants batching efforts.

Reply:
I tried on a JN site with 100k orders created using smooth generator (orders authoritative, sync off, so 100k placeholders were created). I ran `time wp db query "DELETE FROM wp_posts WHERE post_type = 'shop_order_placehold' AND ID IN (SELECT id FROM wp_wc_orders)"` (changing the `ID NOT IN` to `ID IN` so that all the existing records would be affected).

The command took 10s to complete, so if the action limit is 30s it seems like yes, we'll need to batch.

Reply:
Implemented in https://github.com/woocommerce/woocommerce/pull/37050/commits/0759984b9fe560ec7018916834b82e900c13fd21